lib/auth.js

```diff
@@ -0,0 +1,74 @@
+'use strict'
+
+const { parse, GraphQLDirective } = require('graphql')
+const { kApplyPolicy, kAuthContext, kAuthDirective, kGetAuthDirectiveAST, kMakeProtectedResolver } = require('./symbols')
+const { MER_AUTH_ERR_FAILED_POLICY_CHECK } = require('./errors')
+
+class Auth {
+  constructor ({ applyPolicy, authContext, authDirective }) {
+    this[kApplyPolicy] = applyPolicy
+    this[kAuthContext] = authContext
+
+    // Check typeof auth directive option and parse accordingly
+    this[kAuthDirective] = typeof authDirective === 'string' ? parse(authDirective).definitions[0] : authDirective
+  }
+
+  [kGetAuthDirectiveAST] (astNode) {
+    if (Array.isArray(astNode.directives) && astNode.directives.length > 0) {
+      const authDirective = astNode.directives.find(directive => {
+        if (this[kAuthDirective] instanceof GraphQLDirective) {
+          return directive.name.value === this[kAuthDirective].name
+        }
+        return directive.name.value === this[kAuthDirective].name.value
+      })
+      if (typeof authDirective !== 'undefined') {
+        return authDirective
+      }
+    }
+    return null
+  }
+
+  [kMakeProtectedResolver] (authDirectiveAST, resolverFn) {
+    return async (parent, args, context, info) => {
+      const result = await this[kApplyPolicy](authDirectiveAST, parent, args, context, info)
```

Comment:
why somebody would return an Error instead of throwing?

Reply:
I wanted to replicate what is possible within a `graphql-js` resolver - something I learnt yesterday was that GraphQL execute handles errors returned in resolvers as well as when they are thrown, e.g.:

```js
const resolvers = {
  Query: {
    add: async (_, obj) => {
      return new Error('kaboom')
    }
  }
}
```

However, I agree with you that just throwing makes sense and keeps things cleaner so I'll remove this handling! :)

Reply:
Oh I didn't know this. Then +1 in doing this as well. Maybe add a comment?

Reply:
Yeah I didn't know about it either! Cool, yep no problem - I'll add a comment! :)